runtime/libia2/exit_callgates_x86_64.S

```diff
@@ -0,0 +1,82 @@
+// Exit compartment is hardcoded to pkey 1 (where libc/ld.so live)
+#define IA2_EXIT_COMPARTMENT_PKEY 1
+
+// PKRU mask for the exit compartment: allow access to pkey 0 and pkey 1 only.
+// PKRU uses 2 bits per protection key (Access Disable, Write Disable).
+// Result: 0xfffffff0 (pkey 0 and 1 accessible, pkeys 2-15 blocked)
+#define IA2_EXIT_PKRU (~((3U << (2 * IA2_EXIT_COMPARTMENT_PKEY)) | 3U))
+
+    .text
+    .p2align 4,,15
+    .globl ia2_callgate_enter
+    .type ia2_callgate_enter,@function
+ia2_callgate_enter:
+    subq $24, %rsp
+
+    xorl %ecx, %ecx
+    xorl %edx, %edx
+    rdpkru
+    movl %eax, 0(%rsp)              # save caller PKRU
+    leaq 24(%rsp), %r9
+    movq %r9, 8(%rsp)               # save caller stack pointer
+
+    movl $IA2_EXIT_COMPARTMENT_PKEY, %edi
+    call ia2_stackptr_for_compartment@PLT
+
+    movq (%rax), %r10               # load exit stack pointer
+    testq %r10, %r10
+    jne 1f
+    addq $24, %rsp
+    ud2
+1:
+    movl 0(%rsp), %r8d              # reload saved PKRU
+    movq 8(%rsp), %r9               # reload saved SP
+    addq $24, %rsp
+
+    movl $IA2_EXIT_PKRU, %r11d
+    movl %r11d, %eax
+    xorl %ecx, %ecx
+    xorl %edx, %edx
+    wrpkru
+
+#ifdef IA2_DEBUG
+    xorl %ecx, %ecx
+    xorl %edx, %edx
+    rdpkru
+    cmpl %r11d, %eax
+    je 2f
+    ud2
+2:
+#endif
+
+    movq %r10, %rsp                 # switch to exit stack
+    movq %r9, %rdx                  # return saved SP
+    movl %r8d, %eax                 # return saved PKRU
+    ret
+
+    .globl ia2_callgate_exit
+    .type ia2_callgate_exit,@function
+ia2_callgate_exit:
+    testq %rsi, %rsi
+    je 3f
+    movq %rsi, %rsp
+3:
+    movl %edi, %eax
+    xorl %ecx, %ecx
+    xorl %edx, %edx
+    wrpkru
+
+#ifdef IA2_DEBUG
+    xorl %ecx, %ecx
+    xorl %edx, %edx
+    rdpkru
+    cmpl %edi, %eax
+    je 4f
+    ud2
+4:
+#endif
+
+    ret
+
+    .size ia2_callgate_enter, .-ia2_callgate_enter
```

Comment:
nit: This needs to be at the end of `ia2_callgate_enter`. In the second arg to the [`.size` directive](https://sourceware.org/binutils/docs/as/Size.html) "`.`" means "the address where the assembler will emit code next" so we calculate the size by subtracting the starting address of `ia2_callgate_enter` from that. That's what `.-ia2_callgate_enter` is doing. Doesn't really matter though since this only affects some ELF debug info and not codegen.